src/pandas_openscm/index_manipulation.py

```diff
@@ -714,3 +714,95 @@ def update_levels_from_other(
     res = pd.MultiIndex(levels=levels, codes=codes, names=names)
 
     return res
+
+
+def set_levels(
+    ini: pd.MultiIndex, levels_to_set: dict[str, Any | Collection[Any]]
+) -> pd.MultiIndex:
+    """
+    Set the levels of a MultiIndex to the provided values
+
+    Parameters
+    ----------
+    ini
+        Input MultiIndex
+
+    levels_to_set
+        Mapping of level names to values to set. If values is of type `Collection`,
+        it must be of the same length as the MultiIndex. If it is not a `Collection`,
+        it will be set to the same value for all levels.
+
+    Returns
+    -------
+    :
+        New MultiIndex with the levels set to the provided values
+
+    Raises
+    ------
+        TypeError
+            If `ini` is not a MultiIndex
+        ValueError
+            If the length of the values is not equal to the length of the index
+    """
+    # TODO mypy says this is unreachable
+    if not isinstance(ini, pd.MultiIndex):
+        msg = f"Expected MultiIndex, got {type(ini)}"
+        raise TypeError(msg)
+
+    df = ini.to_frame(index=False)
```

Comment:
I don't really know why, but calling `to_frame` is crazy expensive. Try and use https://github.com/openscm/gcages/blob/69452181c365c45f83d59edc3189727f651e5655/src/gcages/index_manipulation.py#L18 and https://github.com/openscm/pandas-openscm/blob/bf65fb54f8855bdb6be64f16acfd01a0a34acf6c/src/pandas_openscm/index_manipulation.py#L349 as inspiration for a way to write this without needing to call `to_frame`. We can talk Monday afternoon if it's still not clear (which is very likely going to be the case as this multi-index stuff is super confusing).